src/lib/mina/test/dynamic-call.unit-test.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * Tests that shows we can call a subcontract dynamically based on the address,
+ * as long as its signature matches the signature our contract was compiled against.
+ *
+ * In other words, the exact implementation/constraints of zkApp methods we call are not hard-coded in the caller contract.
+ */
+import {
+  Bool,
+  UInt64,
+  SmartContract,
+  method,
+  PublicKey,
+  Mina,
+} from '../../../index.js';
+
+type Subcontract = SmartContract & {
+  submethod(a: UInt64, b: UInt64): Promise<Bool>;
+};
+
+// two implementations with same signature of the called method, but different provable logic
+
+class SubcontractA extends SmartContract implements Subcontract {
+  @method.returns(Bool)
+  async submethod(a: UInt64, b: UInt64): Promise<Bool> {
+    return a.greaterThan(b);
+  }
+}
+
+class SubcontractB extends SmartContract implements Subcontract {
+  @method.returns(Bool)
+  async submethod(a: UInt64, b: UInt64): Promise<Bool> {
+    return a.mul(b).equals(UInt64.from(42));
+  }
+}
+
+// caller contract that calls the subcontract
+
+class Caller extends SmartContract {
+  @method
+  async call(a: UInt64, b: UInt64, address: PublicKey) {
+    const subcontract = new Caller.Subcontract(address);
+    await subcontract.submethod(a, b);
+  }
+
+  // subcontract to call. this property is changed below
+  // TODO: having to set this property is a hack, it would be nice to
+  static Subcontract: new (...args: any) => Subcontract = SubcontractA;
```

Comment:
Is this meant to be part of the above comment? The sentence just cuts off :sweat_smile:

Reply:
Heh thanks. Got distracted